src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
         {
-            return Parse<TEnum>(value, false);
+            Debug.Assert(typeof(TResult) == enumType || typeof(TResult) == typeof(object));
+            Debug.Assert(enumType.GetEnumUnderlyingType() == typeof(int));
+
+            bool failureIsOverflow = false;
+            if (StartsNumber(value[0]) && Number.TryParseInt32IntegerStyle(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture.NumberFormat, out int intResult, ref failureIsOverflow))
+            {
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, intResult) : Unsafe.As<int, TResult>(ref intResult);
+                return true;
+            }
+
+            if (failureIsOverflow)
+            {
+                if (throwOnFailure) throw new OverflowException(SR.Overflow_Int32);
+            }
+            else if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulong ulongResult))
+            {
+                intResult = (int)ulongResult;
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, intResult) : Unsafe.As<int, TResult>(ref intResult);
+                return true;
+            }
+
+            enumResult = default;
+            return false;
         }
 
-        public static TEnum Parse<TEnum>(string value, bool ignoreCase) where TEnum : struct
+        private static bool TryParseUInt32Enum<TResult>(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out TResult enumResult)
         {
-            EnumResult parseResult = new EnumResult() { canThrow = true };
-            if (TryParseEnum(typeof(TEnum), value, ignoreCase, ref parseResult))
-                return (TEnum)parseResult.parsedEnum;
-            else
-                throw parseResult.GetEnumParseException();
+            Debug.Assert(typeof(TResult) == enumType || typeof(TResult) == typeof(object));
+            Debug.Assert(enumType.GetEnumUnderlyingType() == typeof(uint));
+
+            bool failureIsOverflow = false;
+            if (StartsNumber(value[0]) && Number.TryParseUInt32IntegerStyle(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture.NumberFormat, out uint uintResult, ref failureIsOverflow))
+            {
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, uintResult) : Unsafe.As<uint, TResult>(ref uintResult);
+                return true;
+            }
+
+            if (failureIsOverflow)
+            {
+                if (throwOnFailure) throw new OverflowException(SR.Overflow_UInt32);
+            }
+            else if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulong ulongResult))
```

Comment:
How common is this case? Does it make sense to move this out into a separate method and aggressively inline the current method?

Essentially, can we move everything outside of the first if:
```C#
if (...)
{
   ...
   return true;
}

return SlowPath(..., failureIsOverflow, out enumResult);
```

Ignore if the TryParseByName is common enough.

Reply:
> How common is this case?

Very common.

> Does it make sense to move this out into a separate method and aggressively inline the current method?

No.